pkg/operator/controllerimagepruner.go

```diff
@@ -0,0 +1,423 @@
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+
+	configapiv1 "github.com/openshift/api/config/v1"
+	imageregistryv1 "github.com/openshift/api/imageregistry/v1"
+	regopset "github.com/openshift/client-go/imageregistry/clientset/versioned"
+	regopinformers "github.com/openshift/client-go/imageregistry/informers/externalversions"
+
+	"github.com/openshift/cluster-image-registry-operator/defaults"
+	regopclient "github.com/openshift/cluster-image-registry-operator/pkg/client"
+	"github.com/openshift/cluster-image-registry-operator/pkg/parameters"
+	"github.com/openshift/cluster-image-registry-operator/pkg/resource"
+	"github.com/openshift/cluster-image-registry-operator/pkg/resource/object"
+	"github.com/openshift/cluster-image-registry-operator/pkg/resource/strategy"
+)
+
+const (
+	imagePrunerWorkQueueKey = "imageprunerchanges"
+)
+
+var (
+	defaultPrunerSuspend                    = false
+	defaultPrunerKeepTagRevisions           = 3
+	defaultPrunerSuccessfulJobsHistoryLimit = int32(3)
+	defaultPrunerFailedJobsHistoryLimit     = int32(3)
+)
+
+// NewImagePrunerController returns a controller for openshift image pruner.
+func NewImagePrunerController(kubeconfig *restclient.Config) (*ImagePrunerController, error) {
+	namespace, err := regopclient.GetWatchNamespace()
+	if err != nil {
+		klog.Fatalf("failed to get watch namespace: %s", err)
+	}
+
+	p := Parameters(namespace)
+
+	listers := &regopclient.ImagePrunerControllerListers{}
+	clients := &regopclient.Clients{}
+	c := &ImagePrunerController{
+		kubeconfig: kubeconfig,
+		params:     *p,
+		generator:  resource.NewImagePrunerGenerator(kubeconfig, clients, listers, p),
+		workqueue:  workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), imagePrunerWorkQueueKey),
+		listers:    listers,
+		clients:    clients,
+	}
+
+	// Initial event to bootstrap the pruner if it doesn't exist.
+	c.workqueue.AddRateLimited(imagePrunerWorkQueueKey)
+
+	return c, nil
+}
+
+// ImagePrunerController keeps track of openshift image pruner components.
+type ImagePrunerController struct {
+	kubeconfig *restclient.Config
+	params     parameters.Globals
+	generator  *resource.ImagePrunerGenerator
+	workqueue  workqueue.RateLimitingInterface
+	listers    *regopclient.ImagePrunerControllerListers
+	clients    *regopclient.Clients
+}
+
+func (c *ImagePrunerController) createOrUpdateResources(cr *imageregistryv1.ImagePruner) error {
+	return c.generator.Apply(cr)
+
+}
+
+type ByCreationTimestamp []*batchv1.Job
+
+func (b ByCreationTimestamp) Len() int {
+	return len(b)
+}
+
+func (b ByCreationTimestamp) Swap(i, j int) {
+	b[i], b[j] = b[j], b[i]
+}
+
+func (b ByCreationTimestamp) Less(i, j int) bool {
+	return !b[j].CreationTimestamp.Time.After(b[i].CreationTimestamp.Time)
+}
+
+// Bootstrap registers this operator with OpenShift by creating an appropriate
+// ClusterOperator custom resource. This function also creates the initial
```

Comment:
This function doesn't create ClusterOperator objects.

Reply:
thanks, i thought i got all of those and removed them.